analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +75,103 @@ public override void ExecutionCompleted()
 
-        foreach (var operation in alwaysTrue)
+        foreach (var (operation, block) in alwaysTrue.Select(x => (x.Key, x.Value)))
         {
-            ReportIssue(operation, true);
+            ReportIssue(operation, block, true);
         }
-        foreach (var operation in alwaysFalse)
+        foreach (var (operation, block) in alwaysFalse.Select(x => (x.Key, x.Value)))
         {
-            ReportIssue(operation, false);
+            ReportIssue(operation, block, false);
         }
 
         base.ExecutionCompleted();
     }
 
-    private void ReportIssue(IOperation operation, bool conditionEvaluation)
+    private void ReportIssue(IOperation operation, BasicBlock block, bool conditionIsTrue)
+    {
+        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionIsTrue);
+        var secondaryLocations = SecondaryLocations(block, conditionIsTrue);
+        if (secondaryLocations.Any())
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionIsTrue));
+        }
+    }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        var unreachable = UnreachableOperations(block, conditionIsTrue).Where(x => !IsIgnoredLocation(x.Syntax));
+        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (unreachable.Contains(operation))
+            {
+                currentStart ??= operation;
+                currentEnd = operation;
+            }
+            else
+            {
+                if (currentStart is not null)
+                {
+                    locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+                    currentStart = null;
+                }
+            }
+        }
+        if (currentStart != null)
+        {
+            locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+        }
+        return locations;
+    }
+
+    private IEnumerable<IOperation> UnreachableOperations(BasicBlock block, bool conditionIsTrue)
     {
-        var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionEvaluation);
-        ReportIssue(Rule2589, operation, null, issueMessage);
+        if (block.SuccessorBlocks.Distinct().Count() != 2)
+        {
+            return Enumerable.Empty<IOperation>();
+        }
+        HashSet<BasicBlock> reachable = new() { block };
+        HashSet<BasicBlock> unreachable = new();
+
+        var reachableSuccessor = conditionIsTrue ^ block.ConditionKind == ControlFlowConditionKind.WhenFalse
```

Comment:
Same thing, way easier to digest
```suggestion
        var reachableSuccessor = (block.ConditionKind == ControlFlowConditionKind.WhenTrue) == conditionIsTrue
```